README.md
# Deprecation notice 🔒

**This repository has been archived (read-only) on Nov 21, 2022**. Thanks to everyone who contributed to `lightning-transformers`, we feel it's time to move on.

:hugs: Transformers can **already be easily trained using the Lightning :zap: Trainer**. Here's a recent example from the community: <https://sachinruk.github.io/blog/deep-learning/2022/11/07/t5-for-grammar-correction.html>. Note that there are **no limitations or workarounds**, things just work out of the box.

The `lightning-transformers` repo explored the possibility to provide task-specific modules and pre-baked defaults, at the cost of introducing extra abstractions. In the spirit of keeping ourselves focused, these abstractions are not something we wish to continue supporting.

If you liked `lightning-transformers` and want to continue developing it in the future, feel free to fork the repo and choose another name for the project.

______________________________________________________________________

<div align="center">

<img src="docs/source/_static/images/logo.png" width="500px">

**Flexible components pairing :hugs: Transformers with [Pytorch Lightning](https://github.com/PyTorchLightning/pytorch-lightning) :zap:**

______________________________________________________________________

<p align="center">
  <a href="https://lightning-transformers.readthedocs.io/">Docs</a> •
  <a href="#community">Community</a>
</p>

______________________________________________________________________

</div>

## Installation

```bash
pip install lightning-transformers
```

<details>
<summary>From Source</summary>

```bash
git clone https://github.com/PyTorchLightning/lightning-transformers.git
cd lightning-transformers
pip install .
```

</details>

______________________________________________________________________

## What is Lightning-Transformers

Lightning Transformers provides `LightningModules`, `LightningDataModules` and `Strategies` to use :hugs: Transformers with the [PyTorch Lightning Trainer](https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html).

## Quick Recipes

#### Train [bert-base-cased](https://huggingface.co/bert-base-cased) on the [CARER](https://huggingface.co/datasets/emotion) emotion dataset using the Text Classification task.

```python
import pytorch_lightning as pl
from transformers import AutoTokenizer

from lightning_transformers.task.nlp.text_classification import (
    TextClassificationDataModule,
    TextClassificationTransformer,
)

tokenizer = AutoTokenizer.from_pretrained(
    pretrained_model_name_or_path="bert-base-cased"
)
dm = TextClassificationDataModule(
    batch_size=1,
    dataset_name="emotion",
    max_length=512,
    tokenizer=tokenizer,
)
model = TextClassificationTransformer(
    pretrained_model_name_or_path="bert-base-cased", num_labels=dm.num_classes
)

trainer = pl.Trainer(accelerator="auto", devices="auto", max_epochs=1)

trainer.fit(model, dm)
```

#### Train a pre-trained [mt5-base](https://huggingface.co/google/mt5-base) backbone on the [WMT16](https://huggingface.co/datasets/wmt16) dataset using the Translation task.

```python
import pytorch_lightning as pl
from transformers import AutoTokenizer

from lightning_transformers.task.nlp.translation import (
    TranslationTransformer,
    WMT16TranslationDataModule,
)

tokenizer = AutoTokenizer.from_pretrained(
    pretrained_model_name_or_path="google/mt5-base"
)
model = TranslationTransformer(
    pretrained_model_name_or_path="google/mt5-base",
    n_gram=4,
    smooth=False,
    val_target_max_length=142,
    num_beams=None,
    compute_generate_metrics=True,
)
dm = WMT16TranslationDataModule(
    # WMT translation datasets: ['cs-en', 'de-en', 'fi-en', 'ro-en', 'ru-en', 'tr-en']
    dataset_config_name="ro-en",
    source_language="en",
    target_language="ro",
    max_source_length=128,
    max_target_length=128,
    padding="max_length",
    tokenizer=tokenizer,
)
trainer = pl.Trainer(accelerator="auto", devices="auto", max_epochs=1)

trainer.fit(model, dm)
```

Lightning Transformers supports a bunch of :hugs: tasks and datasets. See the [documentation](https://lightning-transformers.readthedocs.io/en/latest/).

## Billion Parameter Model Support

### Big Model Inference

It's really easy to enable large model support for the pre-built LightningModule :hugs: tasks.

Below is an example to enable automatic model partitioning (across CPU/GPU and even leveraging disk space) to run text generation using a 6B parameter model.

```python
import torch
from accelerate import init_empty_weights
from transformers import AutoTokenizer

from lightning_transformers.task.nlp.language_modeling import (
    LanguageModelingTransformer,
)

with init_empty_weights():
    model = LanguageModelingTransformer(
        pretrained_model_name_or_path="EleutherAI/gpt-j-6B",
        tokenizer=AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B"),
        low_cpu_mem_usage=True,
        device_map="auto",  # automatically partitions the model based on the available hardware.
    )

output = model.generate("Hello, my name is", device=torch.device("cuda"))
print(model.tokenizer.decode(output[0].tolist()))
```

For more information see [Big Transformers Model Inference](https://lightning-transformers.readthedocs.io/en/latest/features/large_model.html).

### Big Model Training with DeepSpeed

Below is an example of how you can train a 6B parameter transformer model using Lightning Transformers and DeepSpeed.

```python
import pytorch_lightning as pl
from transformers import AutoTokenizer

from lightning_transformers.task.nlp.language_modeling import (
    LanguageModelingDataModule,
    LanguageModelingTransformer,
)

tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path="gpt2")

model = LanguageModelingTransformer(
    pretrained_model_name_or_path="EleutherAI/gpt-j-6B",
    tokenizer=AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B"),
    deepspeed_sharding=True,  # defer initialization of the model to shard/load pre-train weights
)

dm = LanguageModelingDataModule(
    batch_size=1,
    dataset_name="wikitext",
    dataset_config_name="wikitext-2-raw-v1",
    tokenizer=tokenizer,
)
trainer = pl.Trainer(
    accelerator="gpu",
    devices="auto",
    strategy="deepspeed_stage_3",
    precision=16,
    max_epochs=1,
)

trainer.fit(model, dm)
```

For more information see [DeepSpeed Training with Big Transformers Models](https://lightning-transformers.readthedocs.io/en/latest/features/large_model_training.html) or the [Model Parallelism](https://pytorch-lightning.readthedocs.io/en/latest/advanced/model_parallel.html#fully-sharded-training) documentation.

## Contribute

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## Community

For help or questions, join our huge community on [Slack](https://www.pytorchlightning.ai/community)!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.2.5] - 2022-11-21

### Fixed

- Fixed loading HF model ([#306](https://github.com/Lightning-AI/lightning-transformers/pull/306))
- Fixed passing config name to `CNNDailyMailSummarizationDataModule` ([#310](https://github.com/Lightning-AI/lightning-transformers/pull/310))
- Fixed move pipeline to `self.device` as default ([#309](https://github.com/Lightning-AI/lightning-transformers/pull/309))


## [0.2.4] - 2022-11-03

### Changed

- Added support for Lightning v1.8.0 ([#297](https://github.com/Lightning-AI/lightning-transformers/pull/297))


## [0.2.3] - 2022-10-08

### Changed

- Use `lightning-utilities` for compatibility ([#292](https://github.com/Lightning-AI/lightning-transformers/pull/292))


## [0.2.2] - 2022-10-07

### Changed

- Added support for Lightning v1.7.0 ([#284](https://github.com/Lightning-AI/lightning-transformers/pull/284))
- Allow generation, fix examples ([#271](https://github.com/Lightning-AI/lightning-transformers/pull/271))

### Fixed

- Fixed `LightningCLI` compatibility ([#288](https://github.com/Lightning-AI/lightning-transformers/pull/288))


## [0.2.1] - 2022-06-28

### Changed

- Simplified Large Model Support/Add Large Model Training  ([#269](https://github.com/Lightning-AI/lightning-transformers/pull/269))
- Refactored the code for model creation ([#268](https://github.com/Lightning-AI/lightning-transformers/pull/268))


## [0.2.0] - 2022-06-23

### Added

- Added big model support ([#263](https://github.com/Lightning-AI/lightning-transformers/pull/263))
- Added support for `Trainer.predict` method ([#261](https://github.com/Lightning-AI/lightning-transformers/pull/261))
- Added ViT Image Classification Support ([#252](https://github.com/Lightning-AI/lightning-transformers/pull/252))
- Allow streaming datasets for the language modeling task ([#256](https://github.com/Lightning-AI/lightning-transformers/pull/256))
- Micro to macro conversion for Classification Metrics ([#255](https://github.com/Lightning-AI/lightning-transformers/pull/255))
- Added ability to set kwargs for pipeline in load_from_checkpoint ([#204](https://github.com/Lightning-AI/lightning-transformers/pull/204))
- Added Masked Language Modeling ([#173](https://github.com/Lightning-AI/lightning-transformers/pull/173))
- Added more schedulers ([#143](https://github.com/Lightning-AI/lightning-transformers/pull/143))

### Changed

- Refactored for new class based approach ([#243](https://github.com/Lightning-AI/lightning-transformers/pull/243))
- Flatten inheritance ([#245](https://github.com/Lightning-AI/lightning-transformers/pull/245))
- Removed configs ([#264](https://github.com/Lightning-AI/lightning-transformers/pull/264))
- Removed config/hydra from repo ([#262](https://github.com/Lightning-AI/lightning-transformers/pull/262))
- Changed `gpus` to `devices` for compatability ([#239](https://github.com/Lightning-AI/lightning-transformers/pull/239))

### Fixed

- Fixed Token Classification ([#265](https://github.com/Lightning-AI/lightning-transformers/pull/265))
- Fixed broken bleu metrics ([#228](https://github.com/Lightning-AI/lightning-transformers/pull/228))
- Fixed a bug preventing a model evaluating on the test split ([#186](https://github.com/Lightning-AI/lightning-transformers/pull/186))
- Reflect default settings changed to using ddp ([#175](https://github.com/Lightning-AI/lightning-transformers/pull/175))


## [0.1.0] - 2022-04-21

- EVERYTHING


examples/translation_wmt.py
import pytorch_lightning as pl
from transformers import AutoTokenizer

from lightning_transformers.task.nlp.translation import TranslationTransformer, WMT16TranslationDataModule

if __name__ == "__main__":
    tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path="t5-base", model_max_length=512)
    model = TranslationTransformer(
        pretrained_model_name_or_path="t5-base",
        n_gram=4,
        smooth=False,
        val_target_max_length=142,
        num_beams=None,
        compute_generate_metrics=True,
    )
    dm = WMT16TranslationDataModule(
        # WMT translation datasets: ['cs-en', 'de-en', 'fi-en', 'ro-en', 'ru-en', 'tr-en']
        dataset_config_name="ro-en",
        source_language="en",
        target_language="ro",
        max_source_length=128,
        max_target_length=128,
        padding="max_length",
        tokenizer=tokenizer,
    )
    trainer = pl.Trainer(accelerator="auto", devices="auto", max_epochs=1)

    trainer.fit(model, dm)
